### ISSUE TITLE
Exception `AttributeError: 'NoneType' object has no attribute 'distributed_tracing'`

### ISSUE THREAD

**Issue by k2mahajan** [newrelic/newrelic-python-agent#19]
**Description**
This issue happened with Tornado. Within async POST handler, if we finish the request with self.finish() before using AsyncHttp client for making a http request, the transaction is 'exited' with settings set to None. But we land up in generate_request_headers with 'transaction' not set to None but transaction.settings None, leading to exception:

```
Traceback (most recent call last):
File "/home/ketan/repo/iot/newrelic/iot-data-manager/services/common/asynchttp.py", line 24, in fetch
async_http_response = await http_client.fetch(url)
File "/home/ketan/.local/lib/python3.6/site-packages/newrelic/hooks/framework_tornado.py", line 284, in wrap_httpclient_fetch
outgoing_headers = trace.generate_request_headers(current_transaction())
File "/home/ketan/.local/lib/python3.6/site-packages/newrelic/api/cat_header_mixin.py", line 92, in generate_request_headers
if settings.distributed_tracing.enabled:
AttributeError: 'NoneType' object has no attribute 'distributed_tracing'

```

**Expected Behavior**
There should not be any exception

**[Troubleshooting](https://discuss.newrelic.com/t/troubleshooting-frameworks/108787) or [NR Diag]
The issue is happening as we are doing CachedPath.__exit__() in the context of tornado.web.RequestHandler.finish(). If there is any further network call in the same POST handler the transaction object is not None, but transaction.settings is None.

**Steps to Reproduce**
I have created a sample app to reproduce this issue in the exact same manner what we faced.

The server (myapp.py):
```
import tornado.ioloop
import tornado.web
from tornado.httpclient import AsyncHTTPClient
import traceback

class App1(tornado.web.RequestHandler):
    async def post(self):
        try:
            print("Hello from App1")
            self.finish()
            http_client = AsyncHTTPClient()
            response = await http_client.fetch('http://127.0.0.1:9002')
            print(response.body.decode())
        except:
            print(traceback.format_exc())

class App2(tornado.web.RequestHandler):
    async def get(self):
        self.write("Hello from App2")
        self.finish()

def make_app1():
    return tornado.web.Application([
        (r"/", App1),
    ])

def make_app2():
    return tornado.web.Application([
        (r"/", App2),
    ])

if __name__ == "__main__":
    app1 = make_app1()
    app1.listen(9001)
    app2 = make_app2()
    app2.listen(9002)
    tornado.ioloop.IOLoop.current().start()
```

Execution:
```
NEW_RELIC_LICENSE_KEY=<my_newrelic_lic_key> NEW_RELIC_APP_NAME=my-app newrelic-admin run-python myapp.py
```

The client (curl command):
```
curl -X POST http://127.0.0.1:9001
curl -X POST http://127.0.0.1:9001
```
Please note that the issue does not surface during the first call. It happens second call onwards.

Server output:
```
Hello from App1
Hello from App2
Hello from App1
Traceback (most recent call last):
  File "myapp.py", line 12, in post
    response = await http_client.fetch('http://127.0.0.1:9002')
  File "/home/ketan/.local/lib/python3.6/site-packages/newrelic/hooks/framework_tornado.py", line 284, in wrap_httpclient_fetch
    outgoing_headers = trace.generate_request_headers(current_transaction())
  File "/home/ketan/.local/lib/python3.6/site-packages/newrelic/api/cat_header_mixin.py", line 92, in generate_request_headers
    if settings.distributed_tracing.enabled:
AttributeError: 'NoneType' object has no attribute 'distributed_tracing'
```
If we move the call self.finish() to line after http_client.fetch call, the server works fine:
```
Hello from App1
Hello from App2
Hello from App1
Hello from App2
```
It also works fine if i start server as a regular python app without newrelic (`python3 myapp.py`)
I hope this helps.
**Your Environment**
Ubuntu 18.04
Python 3.7

**Comment by k2mahajan** [comment]
The fix #18 has been merged. Closing the issue.

**Comment by exeptionerror** [comment]
All Possible solution added [[Solved] AttributeError: ‘NoneType’ object has no attribute ‘something’](https://flutterq.com/attributeerror-nonetype-object-has-no-attribute/)
